implementations/rust/ockam/ockam_api/src/cli_state/storage/identities_repository_sql.rs

```diff
@@ -0,0 +1,304 @@
+use core::str::FromStr;
+
+use sqlx::*;
+
+use crate::cli_state::{IdentitiesRepository, NamedIdentity};
+use ockam::identity::Identifier;
+use ockam_core::async_trait;
+use ockam_core::compat::sync::Arc;
+use ockam_core::Result;
+use ockam_node::database::{FromSqlxError, SqlxDatabase, ToSqlxType, ToVoid};
+
+/// Implementation of `IdentitiesRepository` trait based on an underlying database
+/// using sqlx as its API, and Sqlite as its driver
+#[derive(Clone)]
+pub struct IdentitiesSqlxDatabase {
+    database: Arc<SqlxDatabase>,
+}
+
+impl IdentitiesSqlxDatabase {
+    /// Create a new database
+    pub fn new(database: Arc<SqlxDatabase>) -> Self {
+        debug!("create a repository for identities");
+        Self { database }
+    }
+
+    /// Create a new in-memory database
+    pub async fn create() -> Result<Arc<Self>> {
+        Ok(Arc::new(Self::new(
+            SqlxDatabase::in_memory("identities").await?,
+        )))
+    }
+}
+
+#[async_trait]
+impl IdentitiesRepository for IdentitiesSqlxDatabase {
+    async fn store_named_identity(
+        &self,
+        identifier: &Identifier,
+        name: &str,
+        vault_name: &str,
+    ) -> Result<NamedIdentity> {
+        let transaction = self.database.begin().await.into_core()?;
+        let is_already_default = self
+            .get_default_named_identity()
+            .await?
+            .map(|n| n.name() == *name)
+            .unwrap_or(false);
+
+        let query = query("INSERT OR REPLACE INTO named_identity VALUES (?, ?, ?, ?)")
+            .bind(identifier.to_sql())
+            .bind(name.to_sql())
+            .bind(vault_name.to_sql())
+            .bind(is_already_default.to_sql());
+        query.execute(&self.database.pool).await.void()?;
+
+        transaction.commit().await.void()?;
+
+        Ok(NamedIdentity::new(
+            identifier.clone(),
+            name.to_string(),
+            vault_name.to_string(),
+            is_already_default,
+        ))
+    }
+
+    async fn delete_identity(&self, name: &str) -> Result<Option<Identifier>> {
+        let named_identity = self.get_named_identity(name).await?;
+        let is_default = named_identity
+            .clone()
+            .map(|i| i.is_default())
+            .unwrap_or(false);
+        let query = query("DELETE FROM named_identity WHERE name=?").bind(name.to_sql());
```

Comment:
Both can be used. I have indeed not been very consistent.

Reply:
Now, I remember think about it along the lines of "when there are more than 5 or so parameters having indices can be useful to quickly count them or check their position". And otherwise use `?`